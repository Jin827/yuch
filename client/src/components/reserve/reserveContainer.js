import React, { Component } from 'react';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';
import compose from 'recompose/compose';
/* --- Components --- */
import Loader from '../../shared/loader';
import * as moment from '../../shared/moment';
/* --- Actions --- */
import { reserve, resetReserve } from '../../actions/reserveAction';
import { showModalAction, hideModalAction } from '../../actions/modalAction';

/* react/no-unused-state: false */
const SimpleModal = Loader({
  loader: () =>
    import('../../shared/simpleModal' /* webpackChunkName: 'simpleModal' */),
});
const SwitchReserve = Loader({
  loader: () =>
    import('./switchReserve' /* webpackChunkName: 'switchReserve' */),
});

const styles = theme => ({
  bigButton: {
    color: '#ffffff',
    margin: theme.spacing.unit,
    width: '9em',
    paddingTop: '5px',
    paddingBottom: '5px',
    [theme.breakpoints.up('md')]: {
      fontSize: '1em',
    },
  },
});
class ReserveContainer extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);

    this.state = {
      submitBtnClicked: false,
      name: '',
      contact: '(0  )    -    ',
      number: '',
      place: '',
      date: moment.inThreeDays,
      time: '12:30',
    };
  }

  handleChange = ({ target: { id, value } }) => {
    this.setState({ [id]: value });
  };

  handleSubmit = async ev => {
    await ev.preventDefault();
    await this.setState({ submitBtnClicked: true });

    const { onReserve } = this.props;
    const { name, contact, number, place, date, time } = this.state;
    const reserveInfo = {
      name,
      contact,
      number,
      place,
      date,
      time,
      createdAt: moment.timeStamp,
    };
    if (
      name === '' ||
      contact === '' ||
      contact === '(0  )    -    ' ||
      !!(contact[11].indexOf('_') !== -1) ||
      !!(contact[12].indexOf('_') !== -1) ||
      !!(contact[13].indexOf('_') !== -1) ||
      number === '' ||
      place === '' ||
      date === '' ||
      time === ''
    ) {
      return null;
    }
    return onReserve(reserveInfo);
  };

  render() {
    const { apiRequest, showModal, classes } = this.props;
    const { submitBtnClicked } = this.state;
    const { inThreeDays } = moment;

    return (
      <div id="reserve">
        <div className="tc white reserve-container">
          <h3 className="white f-en b">Reservation</h3>
          <p>기업체 각종 행사, 모임 단체 식사 주문받습니다.</p>
          <Button
            onClick={() => this.props.showModalAction()}
            variant="contained"
            color="secondary"
            className={`btn--reserve-modal ${classes.bigButton}`}
          >
            예약하기
          </Button>
        </div>
        {showModal && (
          <SimpleModal
            component={
              <SwitchReserve
                apiRequest={apiRequest}
                reserveInfo={this.state}
                inThreeDays={inThreeDays}
                submitBtnClicked={submitBtnClicked}
                handleChange={this.handleChange}
                handleSubmit={this.handleSubmit}
                handleClose={() => this.props.hideModalAction()}
              />
            }
          />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  apiRequest: state.reserve.apiRequest,
  showModal: state.modal.show,
});

const mapDispatchToProps = dispatch => ({
  onReserve: reserveInfo => dispatch(reserve(reserveInfo)),
  onResetReserve: () => dispatch(resetReserve()),
  showModalAction: () => dispatch(showModalAction()),
  hideModalAction: () => {
    dispatch(hideModalAction());
    dispatch(resetReserve());
  },
});

export const Unwrapped = ReserveContainer;
export default compose(
  withStyles(styles),
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
)(ReserveContainer);
