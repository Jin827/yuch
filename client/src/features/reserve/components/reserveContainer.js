import React, { Component } from 'react';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';
import compose from 'recompose/compose';
/* --- Components --- */
import Loader from '../../../shared/loader';
import * as moment from '../../../shared/moment';
/* --- Actions --- */
import { reserve, resetReserve } from '../reserveAction';

/* react/no-unused-state: false */
const SimpleModal = Loader({
  loader: () =>
    import('../../../shared/simpleModal' /* webpackChunkName: 'simpleModal' */),
});
const SwitchReserve = Loader({
  loader: () =>
    import('./switchReserve' /* webpackChunkName: 'switchReserve' */),
});

const styles = theme => ({
  bigButton: {
    margin: theme.spacing.unit,
    width: '9em',
    paddingTop: '5px',
    paddingBottom: '5px',
    [theme.breakpoints.up('md')]: {
      fontSize: '1em',
    },
  },
});
class ReserveContainer extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);

    this.state = {
      show: false,
      submitBtnClicked: false,
      name: '',
      contact: '(0  )    -    ',
      number: '',
      place: '',
      date: moment.inThreeDays,
      time: '12:30',
    };
  }

  handleOpen = () => this.setState({ show: true });

  handleClose = () => {
    this.setState({ show: false });
    this.props.onResetReserve();
  };

  handleChange = ({ target: { id, value } }) => {
    this.setState({ [id]: value });
  };

  handleSubmit = async ev => {
    await ev.preventDefault();
    await this.setState({ submitBtnClicked: true });

    const { onReserve } = this.props;
    const { name, contact, number, place, date, time } = this.state;
    const reserveInfo = {
      name,
      contact,
      number,
      place,
      date,
      time,
      createdAt: moment.timeStamp,
    };
    if (
      name === '' ||
      contact === '' ||
      contact === '(0  )    -    ' ||
      !!(contact[11].indexOf('_') !== -1) ||
      !!(contact[12].indexOf('_') !== -1) ||
      !!(contact[13].indexOf('_') !== -1) ||
      number === '' ||
      place === '' ||
      date === '' ||
      time === ''
    ) {
      return null;
    }
    return onReserve(reserveInfo);
  };

  render() {
    const { apiRequest, classes } = this.props;
    const { show, submitBtnClicked } = this.state;
    const { inThreeDays } = moment;

    return (
      <div id="reserve">
        <div className="tc white reserve-container">
          <h3 className="white f-en b">Reservation</h3>
          <p>기업체 각종 행사, 모임 단체 식사 주문받습니다.</p>
          <Button
            onClick={ev => this.handleOpen(ev)}
            variant="contained"
            color="secondary"
            className={`btn--reserve-modal ${classes.bigButton}`}
          >
            예약하기
          </Button>
        </div>
        {show && (
          <SimpleModal
            show={show}
            handleClose={this.handleClose}
            component={
              <SwitchReserve
                apiRequest={apiRequest}
                reserveInfo={this.state}
                inThreeDays={inThreeDays}
                submitBtnClicked={submitBtnClicked}
                handleChange={this.handleChange}
                handleSubmit={this.handleSubmit}
                handleClose={this.handleClose}
              />
            }
          />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  apiRequest: state.reserve.apiRequest,
});

const mapDispatchToProps = dispatch => ({
  onReserve: reserveInfo => dispatch(reserve(reserveInfo)),
  onResetReserve: () => dispatch(resetReserve()),
});

export const Unwrapped = ReserveContainer;
export default compose(
  withStyles(styles),
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
)(ReserveContainer);
