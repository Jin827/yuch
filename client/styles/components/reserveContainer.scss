.reserve-container {
  margin: 3em 0;
  padding: 2.2em;
  background-color: $c-point1;

  h3 {
    font-size: 1.75em;
    margin-bottom: .5em;
  }

  p {
    margin-bottom: 1.5em;
  }
}

.reserve-btn {
  font-weight: bold;
  background-color: white;
  border: 2px solid $c-point1;
  box-shadow: .2px .2px #866586d0;
  color: $c-point1;
  padding: 7px 17px;
  font-size: .9em;
  letter-spacing: 1.1px;

  @include for-desktop-up {
    font-size: .75em;
    padding: 9.5px 25px;
  }
}

.reserve-btn:hover {
  border: 2px solid $c-point2;
  color: #814981;
  transition: font-size .2s;
  -webkit-transition: font-size .2s;
  -moz-transition: font-size .2s;
  -ms-transition: font-size .2s;
  -o-transition: border .2s color .2s;
}

/* --- Modal --- */
.modal {
  width: 17em;
  padding: 1.7em 1.2em;
  position: absolute;
  box-shadow: 0px 3px 5px -1px rgba(0, 0, 0, 0.2), 0px 5px 8px 0px rgba(0, 0, 0, 0.14), 0px 1px 14px 0px rgba(0, 0, 0, 0.12);
  background-color: #fff;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  -o-transform: translate(-50%, -50%);

  @include for-tablet-portrait-up {
    width: 23em;
    padding: 3em;
  }

  p {
    font-size: .835em;
    margin-bottom: 1.5em;
  }

  .reserve-btn:first-child {
    margin-right: 1.5em;
  }
}

/* --- Form --- */
.reserve-form--box {
  width: 15em;
  margin: 1.2em auto;
}
