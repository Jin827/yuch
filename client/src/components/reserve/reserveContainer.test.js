import React from 'react';
import { shallow } from 'enzyme';
import { Unwrapped as UnwrappedReserveContainer } from './reserveContainer';

describe('<ReserveContainer />', () => {
  describe('renders initial reserve section correctly', () => {
    const showModalAction = jest.fn();
    const props = {
      show: false,
      classes: {
        bigButton: {},
      },
      showModalAction,
    };

    const wrapper = shallow(<UnwrappedReserveContainer {...props} />);

    it('renders correctly', () => {
      expect(wrapper.find('WithStyles(Button)').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    describe('when clicking reserve button', () => {
      beforeEach(() => {
        wrapper.find('.btn--reserve-modal').simulate('click');
        wrapper.setProps({ showModal: true });
      });

      afterEach(() => {
        wrapper.setProps({ showModal: false });
      });

      it('showModalAction called', () => {
        expect(showModalAction).toHaveBeenCalled();
      });

      it('load a reserve Modal component', () => {
        expect(wrapper.find('LoadableComponent').exists()).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
