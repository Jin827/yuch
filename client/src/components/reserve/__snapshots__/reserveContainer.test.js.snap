// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReserveContainer /> renders initial reserve section correctly renders correctly 1`] = `
<div
  id="reserve"
>
  <div
    className="tc white reserve-container"
  >
    <h3
      className="white f-en b"
    >
      Reservation
    </h3>
    <p>
      기업체 각종 행사, 모임 단체 식사 주문받습니다.
    </p>
    <WithStyles(Button)
      className="btn--reserve-modal [object Object]"
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      예약하기
    </WithStyles(Button)>
  </div>
</div>
`;

exports[`<ReserveContainer /> renders initial reserve section correctly when clicking reserve button load a reserve Modal component 1`] = `
<div
  id="reserve"
>
  <div
    className="tc white reserve-container"
  >
    <h3
      className="white f-en b"
    >
      Reservation
    </h3>
    <p>
      기업체 각종 행사, 모임 단체 식사 주문받습니다.
    </p>
    <WithStyles(Button)
      className="btn--reserve-modal [object Object]"
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      예약하기
    </WithStyles(Button)>
  </div>
  <LoadableComponent
    component={
      <LoadableComponent
        handleChange={[Function]}
        handleClose={[Function]}
        handleSubmit={[Function]}
        inThreeDays="1970-01-04"
        reserveInfo={
          Object {
            "contact": "(0  )    -    ",
            "date": "1970-01-04",
            "name": "",
            "number": "",
            "place": "",
            "submitBtnClicked": false,
            "time": "12:30",
          }
        }
        submitBtnClicked={false}
      />
    }
  />
</div>
`;
